### ISSUE TITLE
Web STOMP hangs after a while

### ISSUE THREAD

**Issue by kekekeks** [rabbitmq/rabbitmq-web-stomp#5]
I see a lot of messages like this
```
=ERROR REPORT==== 9-Feb-2014::07:24:04 ===
** Generic server <0.30556.119> terminating 
** Last message in was session_timeout
** When Server state == {session,undefined,
                            {[],[]},
                            undefined,#Ref<0.0.112.43426>,300000,undefined,
                            25000,connecting,undefined,
                            #Fun<rabbit_ws_sockjs.1.120271989>,{},
                            {sockjs_session,
                                {<0.30556.119>,
                                 [{peername,{{127,0,0,1},49363}},
                                  {sockname,{{127,0,0,1},8080}},
                                  {path,"/stomp/276/x1z4pu8e/websocket"},
                                  {headers,[]}]}}}
** Reason for termination == 
** {function_clause,[{gen_server,cast,
                                 [{},sockjs_closed],
                                 [{file,"gen_server.erl"},{line,194}]},
                     {rabbit_ws_sockjs,service_stomp,3,[]},
                     {sockjs_session,emit,2,[]},
                     {sockjs_session,terminate,2,[]},
                     {gen_server,terminate,6,
                                 [{file,"gen_server.erl"},{line,719}]},
                     {proc_lib,init_p_do_apply,3,
                               [{file,"proc_lib.erl"},{line,239}]}]}
```
in RabbitMQ log and then even stomp/info stops responding to HTTP requests. It happens after several hours under load of about 100-150 connections.

RabbitMQ itself and regular STOMP (that one on port 61613) continue working without any issues.

Ubuntu 13.10, RabbitMQ 3.2.3, Erlang R16B03

**Comment by michaelklishin** [comment]
This may or may not still be relevant after #16 (which also upgraded SockJS) but our timeout handling is worth looking into.

**Comment by sega-yarkin** [comment]
We have the same issue sometimes on 3.5.3.
Example, today with 8k active websocket connections and near 5k msgs/s to users, all websocket conns were closed and new connection returned HTTP/400, but at the same time stomp/info returned HTTP/200. Some users used XHR (near 450-500 rps) and they too started to receive errors (HTTP/500).

**Comment by michaelklishin** [comment]
@sega-yarkin anything suspicous in the broker log files?

**Comment by sega-yarkin** [comment]
@michaelklishin many messages like in issue's description only

**Comment by michaelklishin** [comment]
@sega-yarkin thanks.

@essen this seems like a lack of SockJS session timeout handling. Worth looking into for 3.5.5, scheduled for September. Not sure if we can back port the recent changes in master but it's debatable.

**Comment by michaelklishin** [comment]
@sega-yarkin we'll see what we can do for 3.5.5. There were non-trivial changes in web-stomp in 3.6.0 as far as the build system goes which also bump the minimum required Erlang version, so it's quite
unlikely that we can back port them. But chances are this is not Cowboy version-specific.

**Comment by essen** [comment]
I wouldn't be so sure.

I can see what produces this weird crash but it doesn't look like the actual issue. The crash seems to be that session_timeout is triggered *before* the connection is initialized, at which point there is a pid yet, only this weird {}.

To find the real issue we have to find why session_timeout is triggered before the connection is initialized and why this seems to bring down everything with it. It may be something else entirely, like another vital process dying and not restarting. Either way it's in sockjs-erlang for sure. I don't know if it's still on master, but considering the logs it probably still is an issue.

I am wondering if perhaps when it starts to hang there are one or more different error logs?

**Comment by sega-yarkin** [comment]
@essen I found 3 different kinds of error:
1.
```
=ERROR REPORT==== 17-Aug-2015::11:59:04 ===
** Generic server <0.32048.929> terminating
** Last message in was session_timeout
** When Server state == {session,undefined,
                         {[],[]},
                         undefined,#Ref<0.0.17575.5482>,300000,undefined,
                         25000,connecting,undefined,
                         #Fun<rabbit_ws_sockjs.1.120271989>,{},
                         {sockjs_session,
                          {<0.32048.929>,
                           [{peername,{{127,0,0,1},39861}},
                            {sockname,{{127,0,0,1},15674}},
                            {path,"/stomp/743/k7cxiawz/websocket"},
                            {headers,[]}]}}}
** Reason for termination ==
** {function_clause,[{gen_server,cast,
                                 [{},sockjs_closed],
                                 [{file,"gen_server.erl"},{line,196}]},
                     {rabbit_ws_sockjs,service_stomp,3,[]},
                     {sockjs_session,emit,2,[]},
                     {sockjs_session,terminate,2,[]},
                     {gen_server,try_terminate,3,
                                 [{file,"gen_server.erl"},{line,621}]},
                     {gen_server,terminate,7,
                                 [{file,"gen_server.erl"},{line,787}]},
                     {proc_lib,init_p_do_apply,3,
                               [{file,"proc_lib.erl"},{line,237}]}]}
```
2.
```
=ERROR REPORT==== 17-Aug-2015::12:01:52 ===
** Handler sockjs_cowboy_handler terminating in handle/2
   for the reason exit:{normal,
                           {gen_server,call,
                               [<0.27708.868>,
                                {reply,<0.17720.968>,true},
                                infinity]}}
** Options were {service,"/stomp",#Fun<rabbit_ws_sockjs.1.120271989>,{},
                         "http://cdn.sockjs.org/sockjs-0.2.js",false,true,
                         5000,25000,131072,#Fun<rabbit_ws_sockjs.0.116551930>,
                         undefined}
** Handler state was {service,"/stomp",#Fun<rabbit_ws_sockjs.1.120271989>,{},
                              "http://cdn.sockjs.org/sockjs-0.2.js",false,
                              true,5000,25000,131072,
                              #Fun<rabbit_ws_sockjs.0.116551930>,undefined}
** Request was {http_req,#Port<0.25811536>,cowboy_tcp_transport,keepalive,
                         <0.17720.968>,'POST',
                         {1,1},
                         undefined,
                         [<<"stomp_xhr_upstream">>],
                         undefined,<<"stomp_xhr_upstream">>,80,
                         [<<"stomp">>,<<"361">>,
                          <<"6cf68e99-54ce-4e0d-a68f-1df6ce4815b5">>,
                          <<"xhr">>],
                         [<<"361">>,
                          <<"6cf68e99-54ce-4e0d-a68f-1df6ce4815b5">>,
                          <<"xhr">>],
                         <<"/stomp/361/6cf68e99-54ce-4e0d-a68f-1df6ce4815b5/xhr">>,
                         undefined,<<>>,[],
                         [],
                         [],undefined,[],waiting,<<>>,waiting,[],<<>>,
                         {#Fun<cowboy_http.urldecode.2>,crash}}
** Stacktrace: [{gen_server,call,3,[{file,"gen_server.erl"},{line,190}]},
                {sockjs_action,reply_loop,5,[]},
                {sockjs_cowboy_handler,handle,2,[]},
                {cowboy_http_protocol,handler_handle,3,[]}]
```
3.
```
=ERROR REPORT==== 17-Aug-2015::12:09:20 ===
** Generic server <0.787.828> terminating
** Last message in was {'EXIT',<0.52.0>,shutdown}
** When Server state == {session,undefined,
                         {[],[]},
                         undefined,#Ref<0.0.17702.25515>,300000,undefined,
                         25000,connecting,undefined,
                         #Fun<rabbit_ws_sockjs.1.120271989>,{},
                         {sockjs_session,
                          {<0.787.828>,
                           [{peername,{{127,0,0,1},37236}},
                            {sockname,{{127,0,0,1},15674}},
                            {path,"/stomp/417/9m3ckha6/websocket"},
                            {headers,
                             []}]}}}
** Reason for termination ==
** {function_clause,[{gen_server,cast,
                                 [{},sockjs_closed],
                                 [{file,"gen_server.erl"},{line,196}]},
                     {rabbit_ws_sockjs,service_stomp,3,[]},
                     {sockjs_session,emit,2,[]},
                     {sockjs_session,terminate,2,[]},
                     {gen_server,try_terminate,3,
                                 [{file,"gen_server.erl"},{line,621}]},
                     {gen_server,terminate,7,
                                 [{file,"gen_server.erl"},{line,787}]},
                     {proc_lib,init_p_do_apply,3,
                               [{file,"proc_lib.erl"},{line,237}]}]}
```

**Comment by essen** [comment]
Thanks.

Nothing obvious in these logs, and nothing obvious I can see in the sockjs-erlang code either.

@michaelklishin Curious about what the "detailed explanation" mentioned in https://github.com/rabbitmq/rabbitmq-web-stomp/issues/5#issuecomment-34574183 is...

**Comment by michaelklishin** [comment]
We will try stress testing master with a concurrency testing and verification tool(s) but so far cannot reproduce => clearing milestone.

**Comment by michaelklishin** [comment]
web-stomp got a *major* revamp in 3.6.0, so chances are, this is no longer relevant. Closing.
